utf8/reader.go

```diff
@@ -0,0 +1,142 @@
+package utf8
+
+import (
+	"fmt"
+	"io"
+	"unicode/utf8"
+
+	"github.com/madlambda/spells/io/bytes"
+)
+
+type (
+	RuneReader struct {
+		r io.RuneReader
+	}
+
+	// ReaderReader implements a runes.Reader from a bytes.Reader interface.
+	ReaderReader struct {
+		r bytes.Reader
+	}
+
+	Error string
+)
+
+const ErrDecode Error = "invalid rune"
+
+// NewReaderReader creates a runes.Reader implementation from an bytes.Reader
+// interface.
+func NewReaderReader(r bytes.Reader) *ReaderReader {
```

Comment:
I also suggested at least calling it Reader instead of ReaderReader...but now thinking on the type itself, and what it does, I'm thinking more on terms of a decoder, it decodes a byte stream of utf-8 to runes, which is also an implementation of the runes.Reader interface, but that doesn't imply the name should be Reader too, it is not just reading utf-8 or unicode, it is actively decoding.

```suggestion
func NewDecoder(r bytes.Reader) *Decoder {
```